### PR TITLE
feat(che-theia): Resolve plug-ins for che-theia editor

### PR DESCRIPTION
**Pull request by benoitf**
### What does this PR do?
Resolve che-theia plug-ins
It embeds the che-theia and devfile library to update devWorkspace templates if there are some che-theia plug-ins
It also resolve optional files content

https://user-images.githubusercontent.com/436777/121185653-9364ac80-c866-11eb-8c60-813f4fc74657.mp4

### What issues does this PR fix or reference?
https://github.com/eclipse/che/issues/19927

### how to test PR 
1. Use sample from get started page

OR

2. Use $CHE_HOST/#https://github.com/che-samples/java-spring-petclinic/blob/devfilev2/devfile.yaml as factory link and make sure redhat/java plugin is present after workspace is started

### CODE REVIEW

**Comment by benoitf** [comment]
[dashboard-ci-test]

**Review by sleshchenko — COMMENTED** [review]
I can't say anything about code, but I tested and it seems to work as expected. Oleksiis are reviewing changes, let's wait them

File: src/services/workspace-client/devWorkspaceClient.ts [hunk 1]
```diff
@@ -127,10 +133,54 @@ export class DevWorkspaceClient extends WorkspaceClient {
...
+      if (!template.metadata?.['namespace']) {
+        const templateMetadata: { namespace?: string } = template.metadata || {};
+        if (!templateMetadata.namespace) {
+          templateMetadata.namespace = namespace;
+        }
+        template.metadata = templateMetadata;
+      }
+      const theiaDWT = await this.dwtApi.create(<IDevWorkspaceTemplate>template);
+      const pluginName = theiaDWT.metadata.name;
+      this.addPlugin(createdWorkspace, pluginName, theiaDWT.metadata.namespace);
```
Comment (sleshchenko) [inline src/services/workspace-client/devWorkspaceClient.ts:182]: for some reason vsxinstaller did not get ownerRef and is left on the cluster 
Screenshot_20210609_141744

**Comment by sleshchenko** [comment]
rebased against main branch and solved conflicts. You can check that I haven't modified PR code, but just adapt Josh's changes https://github.com/eclipse-che/che-dashboard/compare/7fb464f80ba1c0145beb3b2529a5b6d5d08a8a79..c48a9bf92183845a900b347129991a595bfcb134

File: src/services/workspace-client/devWorkspaceClient.ts [hunk 2]
```diff
@@ -149,10 +155,54 @@ export class DevWorkspaceClient extends WorkspaceClient {
...
+      cheTheiaPluginsContent,
+      vscodeExtensionsJsonContent,
+      devWorkspace,
+      devWorkspaceTemplates,
+      sidecarPolicy,
+      suffix: workspaceId,
+    });
+
+    await Promise.all(devWorkspaceTemplates.map(async template => {
+      // if no namespace, update it
```
Comment (sleshchenko) [inline src/services/workspace-client/devWorkspaceClient.ts:194]: Should we really try to create dwt in the namespace different than devworkspace one?

File: src/services/workspace-client/devWorkspaceClient.ts (hunk 1 above)
Comment (sleshchenko) [reply src/services/workspace-client/devWorkspaceClient.ts:204]: I'm preparing the fix and going to push it after testing

Comment (sleshchenko) [reply src/services/workspace-client/devWorkspaceClient.ts]: fixed with https://github.com/eclipse-che/che-dashboard/pull/266/commits/b8e30b469052c210c61ab5228e104df6a8d1b582

File: src/services/workspace-client/devWorkspaceClient.ts (hunk 2 above)
Comment (benoitf) [reply src/services/workspace-client/devWorkspaceClient.ts:190]: probably not

File: src/pages/GetStarted/CustomWorkspaceTab/index.tsx [hunk 3]
```diff
@@ -162,7 +163,7 @@ export class CustomWorkspaceTab extends React.PureComponent<Props, State> {
       isCreated: true
     });
     try {
-      await this.props.onDevfile(this.state.devfile, this.state.namespace?.name);
+      await this.props.onDevfile(this.state, this.state.namespace?.name);
```
[the remark this answers is not in the document]
Comment (benoitf) [reply src/pages/GetStarted/CustomWorkspaceTab/index.tsx:166]: we need more stuff than only the devfile this is why I was passing the full object. If it's bad then I would need to pass all entries of this.state

File: src/pages/GetStarted/GetStartedTab/SamplesListGallery.tsx [hunk 4]
```diff
@@ -39,7 +39,9 @@ import stringify from '../../../services/helpers/editor';
 type Props =
   MappedProps
   & {
-    onCardClick: (devfileContent: string, stackName: string) => void;
+    onCardClick: (devfileContent: string, stackName: string, optionalFilesContent: {
```
[the remark this answers is not in the document]
Comment (benoitf) [reply src/pages/GetStarted/GetStartedTab/SamplesListGallery.tsx:42]: ok

File: src/store/FactoryResolver.ts [hunk 5]
```diff
@@ -56,6 +64,14 @@ interface ReceiveFactoryResolverAction {
     location?: string;
     source?: string;
     devfile?: api.che.workspace.devfile.Devfile;
+    scm_info: {
```
[the remark this answers is not in the document]
Comment (benoitf) [reply src/store/FactoryResolver.ts:67]: ok

File: src/services/workspace-client/devWorkspaceClient.ts (hunk 2 above)
Comment (sleshchenko) [reply src/services/workspace-client/devWorkspaceClient.ts:190]: then let's remote if statement and always override namespace.

File: src/pages/GetStarted/CustomWorkspaceTab/index.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (akurinnoy) [reply src/pages/GetStarted/CustomWorkspaceTab/index.tsx:166]: I mean that `onDevfile` handler expects ResolverState but takes the CustomWorkspaceTab component state. These interfaces both have the `devfile` field in common but `optionalFilesContent` is missed in the component state. 

`resolverState.optionalFileContent` will always be undefined here:
https://github.com/eclipse-che/che-dashboard/blob/b8e30b469052c210c61ab5228e104df6a8d1b582/src/pages/GetStarted/index.tsx#L157

[the remark this answers is not in the document]
Comment (benoitf) [reply src/pages/GetStarted/CustomWorkspaceTab/index.tsx:166]: hi, would it be more convenient to you to commit to the branch directly ?

[the remark this answers is not in the document]
Comment (akurinnoy) [reply src/pages/GetStarted/CustomWorkspaceTab/index.tsx:166]: yes, sure

File: src/pages/GetStarted/GetStartedTab/SamplesListGallery.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (benoitf) [reply src/pages/GetStarted/GetStartedTab/SamplesListGallery.tsx]: fixd

File: src/services/workspace-client/devWorkspaceClient.ts (hunk 2 above)
Comment (benoitf) [reply src/services/workspace-client/devWorkspaceClient.ts]: fixed

File: src/store/FactoryResolver.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (benoitf) [reply src/store/FactoryResolver.ts]: fixed

**Review by akurinnoy — APPROVED** [review]

**Review by sleshchenko — APPROVED** [review]

**Review by olexii4 — APPROVED** [review]

**Comment by sleshchenko** [comment]
I'm waiting to happy path test to pass and will merge this PR

**Pull request merged**